File: individual/client/src/main/java/requests/DataPacket.java
package requests;

import fields.OpCode;

import java.nio.ByteBuffer;

public class DataPacket extends Packet {

    private final short blockNumber;

    private final String data;

    public DataPacket(short blockNumber, String data) {
        super(OpCode.DATA);
        this.blockNumber = blockNumber;
        this.data = data;
    }

    @Override
    public int size() {
        return Character.BYTES  + Character.BYTES  + data.length();
    }

    @Override
    public byte[] serialize() {
        ByteBuffer buf = ByteBuffer.allocate(size());
        buf.putShort(getOpCode());
        buf.putChar((char)blockNumber);
        buf.put(data.getBytes());
        return buf.array();
    }
}

File: udp/src/main/kotlin/ru/hse/anstkras/tftp/packet/DataPacket.kt
package ru.hse.anstkras.tftp.packet

import java.nio.ByteBuffer

class DataPacket(val blockNum: Int, val data: ByteBuffer) :
    Packet {

    //    2 bytes     2 bytes      n bytes
    //    ----------------------------------
    //    | Opcode |   Block #  |   Data     |
    //    ----------------------------------
    @OptIn(ExperimentalUnsignedTypes::class)
    override fun getBytesRepresentation(): ByteBuffer {
        val buffer = ByteBuffer.allocate(4 + data.remaining())
        buffer.putShort(3)
        buffer.putShort(blockNum.toUShort().toShort())
        buffer.put(data)
        return buffer
    }

    companion object : Parsable<Packet> {
        override fun parse(byteBuffer: ByteBuffer): DataPacket {
            val blockNum = byteBuffer.short
            val data = byteBuffer.slice()
            return DataPacket(blockNum.toInt(), data)
        }
    }
}